pkg/didcomm/crypto/legacy/authcrypt/decrypt.go

```diff
@@ -55,84 +49,80 @@ func (c *Crypter) Decrypt(envelope []byte, recipient cryptoutil.KeyPair) ([]byte
+	var candidateKeys []string
+
 	for _, candidate := range recipients {
-		header := candidate.Header
-		pubKey := base58.Decode(header.KID)
-
-		if !bytes.Equal(pubKey, recKey.Pub[:]) {
-			continue
-		}
-
-		pk, err := publicEd25519toCurve25519(recKey.Pub)
-		if err != nil {
-			return nil, err
-		}
-		sk, err := secretEd25519toCurve25519(recKey.Priv)
-		if err != nil {
-			return nil, err
-		}
-
-		sender, err := decodeSender(header.Sender, pk, sk)
-		if err != nil {
-			return nil, err
-		}
-
-		nonceSlice, err := base64.URLEncoding.DecodeString(header.IV)
-		if err != nil {
-			return nil, err
-		}
-		var nonce [24]byte
-		copy(nonce[:], nonceSlice)
-
-		encCEK, err := base64.URLEncoding.DecodeString(candidate.EncryptedKey)
-		if err != nil {
-			return nil, err
-		}
-
-		cekSlice, success := box.Open(nil, encCEK, &nonce, (*[CurveKeySize]byte)(sender), (*[CurveKeySize]byte)(sk))
-		if !success {
-			return nil, errors.New("failed to decrypt CEK")
-		}
-
-		var cek [chacha.KeySize]byte
-		copy(cek[:], cekSlice)
-
-		return &cek, nil
-	}
-
-	return nil, errors.New("no key accessible")
-}
+		candidateKeys = append(candidateKeys, candidate.Header.KID)
+	}
 
-func decodeSender(b64Sender string, pk *publicCurve25519, sk *privateCurve25519) (*publicCurve25519, error) {
-	encSender, err := base64.URLEncoding.DecodeString(b64Sender)
+	recKeyIdx, err := w.FindVerKey(candidateKeys)
+	if err != nil {
+		return nil, fmt.Errorf("no key accessible %w", err)
+	}
+
+	recip := recipients[recKeyIdx]
+	recKey := recip.Header.KID
+
+	recCurvePub, err := w.ConvertToEncryptionKey(recKey)
+	if err != nil {
+		return nil, err
+	}
+
+	sender, err := decodeSender(recip.Header.Sender, recCurvePub, w)
 	if err != nil {
 		return nil, err
 	}
 
-	senderSlice, err := sodiumBoxSealOpen(encSender, pk, sk)
+	nonceSlice, err := base64.URLEncoding.DecodeString(recip.Header.IV)
 	if err != nil {
 		return nil, err
 	}
 
+	encCEK, err := base64.URLEncoding.DecodeString(recip.EncryptedKey)
+	if err != nil {
+		return nil, err
+	}
+
+	cekSlice, err := w.GetBox().BoxEasyOpen(encCEK, nonceSlice, sender, recCurvePub)
+	if err != nil {
+		return nil, fmt.Errorf("failed to decrypt CEK: %s", err)
+	}
+
+	var cek [chacha.KeySize]byte
+	copy(cek[:], cekSlice)
+
+	return &cek, nil
+}
+
+func decodeSender(b64Sender, pk string, w *wallet.BaseWallet) (string, error) {
+	encSender, err := base64.URLEncoding.DecodeString(b64Sender)
+	if err != nil {
+		return "", err
+	}
+
+	senderSlice, err := w.GetBox().BoxSealOpen(encSender, pk)
+	if err != nil {
+		return "", err
+	}
+
 	senderData := base58.Decode(string(senderSlice))
 
 	var senderEdPub [ed25519.PublicKeySize]byte
 	copy(senderEdPub[:], senderData)
 
-	sender, err := publicEd25519toCurve25519((*publicEd25519)(&senderEdPub))
+	sender, err := cryptoutil.GetEncryptionPubKey(base58.Encode(senderData))
 	if err != nil {
-		return nil, err
+		return "", err
```

Comment:
can be simplified or wrap an error `return cryptoutil.GetEncryptionPubKey(base58.Encode(senderData))`